yield_plugin/src/yield_plugin.cpp

```diff
@@ -403,106 +415,243 @@ namespace yield_plugin
     jmt_trajectory.trajectory_points = jmt_trajectory_points;
     return jmt_trajectory;
   }
+
+  void YieldPlugin::external_objects_callback(const carma_perception_msgs::msg::ExternalObjectList::UniquePtr msg)
+  {
+    external_objects_ = msg->objects;
+  }
   
-  carma_planning_msgs::msg::TrajectoryPlan YieldPlugin::update_traj_for_object(const carma_planning_msgs::msg::TrajectoryPlan& original_tp, double initial_velocity) 
+  boost::optional<rclcpp::Time> YieldPlugin::detect_collision_time(const carma_planning_msgs::msg::TrajectoryPlan& trajectory1, const std::vector<carma_perception_msgs::msg::PredictedState>& trajectory2, double collision_radius) 
   {
-        
+    // Iterate through each pair of consecutive points in the trajectories
+
+    RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Starting a new collision detection, trajectory size: " << trajectory1.trajectory_points.size() << ". prediction size: " << trajectory2.size());
+
+    // Iterate through the object to check if it's on the route
+    bool on_route = false;
+    int on_route_idx = 0;
+
+    for (int j = 0; j < trajectory2.size(); ++j) 
+    {
+      lanelet::BasicPoint2d curr_point;
+      curr_point.x() = trajectory2[j].predicted_position.position.x;
+      curr_point.y() = trajectory2[j].predicted_position.position.y;
+
+      auto corresponding_lanelets = wm_->getLaneletsFromPoint(curr_point, 4); // get 4 lanelets
+
+      for (auto llt: corresponding_lanelets)
+      {
+        RCLCPP_DEBUG_STREAM(rclcpp::get_logger("yield_plugin"), "Checking llt: " << llt.id());
+
+        if (route_llt_ids_.find(llt.id()) != route_llt_ids_.end())
+        {
+          on_route = true;
+          on_route_idx = j;
+          break;
+        }
+      }
+      if (on_route)
+        break;
+    }
```

Comment:
Do we need an indexing `for` loop here? From what I can see, a range-based `for` loop suffices, and it provides bounds safety.

Reply:
Yeah we need the indexing here because of the `on_route_idx` variable. This trajectory is being looped again in the next nested loop, so small performance improvement to start from the index that is already on the route.